### PR TITLE
feat(weave): Support video input and output in SDK

### PR DESCRIPTION
**Pull request by zbirenbaum**
## Description

- Towards WB-22961

This is a scoped down iteration that introduces Video handling to the SDK.
- Image handling has been used as a reference implementation.
- Like the image type handler supports only pillow images, this adds support for moviepy.VideoClip and all subclasses (the video library used by models)

Points to note:
There are *a lot* of edge cases here:
1. Moviepy does not seem particularly stable
2. Attempting to handle more than just VideoFileClips (videos which are loaded from valid files with a known extension) required extra attributes to be set in tests. This may just be a problem with the constructed ones missing metadata that would normally be present
3. I included some fairly extensive test cases, but some strange backend issues with webm encoding still seem to be present
4. Other features like Image preview will come in followup PRs. They were working fine, but there's plenty to review here already and having the load op return a type other than what was directly passed in breaks some of our API promises to the client. The way around this would be to include something like the VideoWrapper in the test cases which the client imports, but @neutralino1 and I discussed avoiding wrapped types so this requires a bit more thought

## Testing
Pretty extensive tests were added. I may cut out some of them, as there's so much going on it got to the point that claude wanted to make a README for the test DIR

### CODE REVIEW

File: tests/trace/type_handlers/Video/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Video Type Handler Tests
```
Comment (neutralino1) [inline tests/trace/type_handlers/Video/README.md:1]: Any reason to capitalize the directory name?

File: weave/type_handlers/Video/__init__.py
Comment (neutralino1) [inline weave/type_handlers/Video/__init__.py:1]: Same here re: directory name capitalization.

File: weave/type_handlers/Video/video.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+else:
+    dependencies_met = True
+
+if TYPE_CHECKING:
+    from moviepy.editor import (
+        VideoClip,
+        VideoFileClip,
+    )
+
+SUPPORTED_FORMATS = ["gif", "mp4", "webm"]
```
Comment (neutralino1) [inline weave/type_handlers/Video/video.py:28]: Maybe use an enum here?

File: weave/type_handlers/Video/video.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        obj: The VideoClip or VideoWithPreview to save
+        artifact: The artifact to save to
+        name: Ignored, see comment below
+    """
+    is_video_file = isinstance(obj, VideoFileClip)
+
+    if is_video_file:
+        video_format = get_format_from_filename(obj.filename) or DEFAULT_VIDEO_FORMAT
+    else:
+        video_format = getattr(obj, "format", DEFAULT_VIDEO_FORMAT)
```
Comment (neutralino1) [inline weave/type_handlers/Video/video.py:68]: I'm not sure I understand the default treatment.
According to your code, if we don't recognize the video format, we assume it is `gif`. So we will write an unknown video format to a GIF file. Do we have any guarantee that playback will work?

**Review by neutralino1 — CHANGES_REQUESTED** [review]
Thank you. A few questions.

File: tests/trace/type_handlers/Video/video_edge_cases_test.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+import weave
+
+"""This module tests edge cases in the Video type handler."""
+
+
+@pytest.fixture
+def sample_mp4_path():
+    """Create a sample MP4 file and return its path."""
+    with tempfile.NamedTemporaryFile(suffix=".mp4", delete=False) as tmp:
```
Comment (andrewtruong) [inline tests/trace/type_handlers/Video/video_edge_cases_test.py:16]: this kinda defeats the purpose of a temporary file

It's better to keep `delete=True` and then `yield tmp_path` below.  Then at the end of the test, the temporary file will be automatically cleaned up

File: weave/type_handlers/Video/video.py (hunk 3 above)
Comment (zbirenbaum) [reply weave/type_handlers/Video/video.py:68]: The write_gif method should do the encoding with ffmpeg behind the scenes. If we think audio will often be a part of the clip we can do MP4 instead

File: tests/trace/type_handlers/Video/README.md (hunk 1 above)
Comment (zbirenbaum) [reply tests/trace/type_handlers/Video/README.md:1]: Convention, all the other type handlers have capital names for their directories. I guess because it refers to a type 'Video' and types are capitalized

File: weave/type_handlers/Video/video.py (hunk 2 above)
Comment (zbirenbaum) [reply weave/type_handlers/Video/video.py]: I made an enum and constructed the supported formats from its values. Should make it easier/clearer when adding or removing formats

File: tests/trace/type_handlers/Video/video_edge_cases_test.py (hunk 4 above)
Comment (zbirenbaum) [reply tests/trace/type_handlers/Video/video_edge_cases_test.py]: Good catch, fixed. Quick question though, what's the functional difference between yielding and returning from the test?

Comment (andrewtruong) [reply tests/trace/type_handlers/Video/video_edge_cases_test.py]: When you `return`, you just get the value.

When you `yield`, you can put cleanup code afterwards.  In this case you don't explicitly have to write any cleanup because the `TemporaryFile` contextmanager already does it for you, so you can just do

```py
with NamedTemporaryFile(...) as tmp:
    clip = ...
    yield tmp.name

# then the file will get cleaned up automatically when it gets here
```

File: tests/trace/type_handlers/Video/video_edge_cases_test.py [hunk 5]
```diff
@@ -22,26 +23,26 @@ def sample_mp4_path():
         tmp_path, codec="libx264", audio=False, verbose=False, logger=None
     )
 
-    return tmp_path
+    yield tmp_path
```
Comment (andrewtruong) [inline tests/trace/type_handlers/Video/video_edge_cases_test.py:26]: you have to yield inside the context manager, otherwise the file will get cleaned up before it gets to the test

File: tests/trace/type_handlers/Video/video_edge_cases_test.py (hunk 4 above)
Comment (zbirenbaum) [reply tests/trace/type_handlers/Video/video_edge_cases_test.py]: Oh now I see what you mean, this would be pretty useful in some other scenarios, thank you for bringing it up

File: tests/trace/type_handlers/Video/video_edge_cases_test.py (hunk 5 above)
Comment (zbirenbaum) [reply tests/trace/type_handlers/Video/video_edge_cases_test.py:26]: I wonder why the tests are still passing with it like this. That might indicate it's not working how we think it is

Comment (zbirenbaum) [reply tests/trace/type_handlers/Video/video_edge_cases_test.py]: Fixed, the file was being written after the tempfile was closed. I put both the writing and the yield inside and now it works properly. Nice catch.

**Review by neutralino1 — APPROVED** [review]
Thank you. One naming nit.

File: weave/type_handlers/Video/video.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    # Assume there can only be 1 video in the artifact
+    for filename in artifact.path_contents:
+        path = artifact.path(filename)
+        if filename.startswith("video."):
+            return VideoFileClip(path)
+
+    raise ValueError("No video or found for artifact")
+
+
+def is_instance(obj: Any) -> bool:
```
[the remark this answers is not in the document]
Comment (zbirenbaum) [reply weave/type_handlers/Video/video.py]: That is a bit clearer, fixed
